### PR TITLE
STEP-1205: Add Relaunch Tests for Each Repetition step input

### PR DESCRIPTION
**Pull request by Bence1001**
### Checklist

- [x] I've read and accepted the [Contribution Guidelines](https://github.com/bitrise-steplib/.github/blob/main/CONTRIBUTING.md)
- Requires _MINOR_ [version update](https://semver.org/)

### Context
**We are adding Xcode 13's Test Repetition features to the Xcode Test step.**
* This includes choosing test repetition mode, setting maximum test repetitions and rerunning tests in a different process.

### Changes
* **New step input: Relaunch Tests for Each Repetition (`relaunch_tests_for_each_repetition`).**
    * Works exactly like `Relaunch Tests for Each Repetition` in Xcode's Test Plan Configuration window.
    * Possible values: `yes` or `no`.
    * Sets the appropriate flag on `xcodebuild`: `-relaunch-tests-for-each-repetition`.
    * Can only be used if `Test Repetition Mode` is other than `none`.
* Added eventually failing in-memory tests in sample app: https://github.com/bitrise-io/sample-swift-project-with-parallel-ui-test/pull/8.
* Added E2E tests for input validation and running eventually failing in-memory tests in separate processes.
* Finalized title & description for the 3 new Test Repetition inputs.

### CODE REVIEW

**Comment by Bence1001** [comment]
**DO NOT MERGE!**

Next steps:
1. Wait for CI to pass.
2. Merge https://github.com/bitrise-io/sample-swift-project-with-parallel-ui-test/pull/7.
3. Update sample app branch to `master` in E2E tests.
4. Merge this PR.

File: e2e/bitrise.yml [hunk 1]
```diff
@@ -825,6 +870,50 @@ workflows:
...
+        run_if: '{{enveq "XCODE_MAJOR_VERSION_AT_LEAST_13" "true"}}'
+        inputs:
+        - workflow_id: utility_test_relaunch_tests_for_each_repetition_succeeds
+        - bitrise_config_path: ./e2e/bitrise.yml
+
+  utility_test_relaunch_tests_for_each_repetition_succeeds:
+    envs:
+    - TEST_APP_NUMBER_OF_INITIAL_TEST_SUCCESSES: 1
+    - TEST_APP_URL: https://github.com/bitrise-io/sample-swift-project-with-parallel-ui-test.git
+    - TEST_APP_BRANCH: STEP-1205-eventually-failing-in-memory-tests
```
Comment (Bence1001) [inline e2e/bitrise.yml:897]: Update this here to `master` if this PR is merged: https://github.com/bitrise-io/sample-swift-project-with-parallel-ui-test/pull/8.

File: step.yml [hunk 2]
```diff
@@ -163,7 +163,7 @@ inputs:
       is_required: true
   - test_repetition_mode: "none"
     opts:
-      title: Test Repetition Mode
+      title: Test Repetition Mode (Available in Xcode 13+)
```
Comment (BanyikAnna) [inline step.yml:166]: I understand you wanted to highlight the availability of this feature in the title but it would be better to put a heads-up on it in the description: "Please note that the Test Repetition Mode is only available in Xcode 13+."

File: step.yml [hunk 3]
```diff
@@ -181,14 +181,25 @@ inputs:
         - "up_until_maximum_repetitions"
   - maximum_test_repetitions: 3
     opts:
-      title: Maximum Test Repetitions
+      title: Maximum Test Repetitions (Available in Xcode 13+)
```
Comment (BanyikAnna) [inline step.yml:184]: same as above

File: step.yml [hunk 4]
```diff
@@ -181,14 +181,25 @@ inputs:
...
       category: Test Repetition
       summary: The maximum number of times a test will repeat based on Test Repetition Mode.
       description: |-
         The maximum number of times a test will repeat based on Test Repetition Mode.
         
         Should be more than 1 if Test Repetition Mode (`test_repetition_mode`) is other than `none`.
       is_required: true
+  - relaunch_tests_for_each_repetition: "no"
+    opts:
+      title: Relaunch Tests for Each Repetition (Available in Xcode 13+)
```
Comment (BanyikAnna) [inline step.yml:194]: same as above but up to you

File: step.yml [hunk 5]
```diff
@@ -261,7 +272,7 @@ inputs:
       category: Debug
       title: "Run xcodebuild test only"
       description: |-
-        If `single_build` is set to false, the step runs `xcodebuild OPTIONS build OPTIONS` before the test 
+        If `single_build` is set to false, the step runs `xcodebuild OPTIONS build OPTIONS` before the test
```
Comment (BanyikAnna) [inline step.yml:275]: step > Step

Comment (BanyikAnna) [reply step.yml:275]: If `single_build` is set to false, the step runs `xcodebuild OPTIONS build OPTIONS` before the test. (period was missing)

File: step.yml [hunk 6]
```diff
@@ -261,7 +272,7 @@ inputs:
...
         to generate the project derived data. After that comes `xcodebuild OPTIONS build test OPTIONS`. This command's log is presented in the step's log.
```
Comment (BanyikAnna) [inline step.yml:276]: After that comes `xcodebuild OPTIONS build test OPTIONS`. >> This is followed by `xcodebuild OPTIONS build test OPTIONS`.

Comment (BanyikAnna) [reply step.yml:276]: step > Step

File: step.yml [hunk 7]
```diff
@@ -261,7 +272,7 @@ inputs:
...
 
         If `single_build` is set to true, then the step calls only `xcodebuild OPTIONS build test OPTIONS`.
```
Comment (BanyikAnna) [inline step.yml:278]: step > Step

File: step.yml [hunk 8]
```diff
@@ -289,8 +300,8 @@ inputs:
       category: Debug
       title: "Should retry test on failure?"
       description: |-
-       If you set this input to `yes`, the Step will rerun your tests. With Xcode 13 and above, only your failed test cases will be rerun. With older Xcode versions, all test cases will be rerun. 
-        Please note that this feature is only available from Xcode 13 and above, earlier versions support rerunning ALL test cases instead of a specific failed one.  
+       If you set this input to `yes`, the Step will rerun your tests. With Xcode 13 and above, only your failed test cases will be rerun. With older Xcode versions, all test cases will be rerun.
+        Please note that this feature is only available from Xcode 13 and above, earlier versions support rerunning ALL test cases instead of a specific failed one.
```
Comment (BanyikAnna) [inline step.yml:304]: correct me if I'm wrong but I think the previous two sentences say exactly the same so I dont think we need these two sentences: "Please note that this feature is only available from Xcode 13 and above, earlier versions support rerunning ALL test cases instead of a specific failed one."

File: step.yml [hunk 9]
```diff
@@ -181,14 +181,25 @@ inputs:
...
   - maximum_test_repetitions: 3
     opts:
-      title: Maximum Test Repetitions
+      title: Maximum Test Repetitions (Available in Xcode 13+)
       category: Test Repetition
       summary: The maximum number of times a test will repeat based on Test Repetition Mode.
       description: |-
         The maximum number of times a test will repeat based on Test Repetition Mode.
         
         Should be more than 1 if Test Repetition Mode (`test_repetition_mode`) is other than `none`.
```
Comment (BanyikAnna) [inline step.yml:190]: .... if the Test Repetition Mode (`test_repetition_mode`) is other than `none`.

File: step.yml (hunk 2 above)
Comment (Bence1001) [reply step.yml:166]: I think there are pros and cons of both approaches. In this case, I'd suggest to keep it in the title for the following reasons:
* Xcode 13 has not even been released yet. When it becomes available (this fall), it will take some time for developers to migrate to. Knowing that something is available only on the newest stack/Xcode version is vital information that I think is worth mentioning in the title.
* Keeping up with the availability of features (both framework and tooling) is already a bit of pain for iOS developers; I think having an overview of what's available to you and what is not just by looking at the titles of the step inputs can be very helpful.
* Since we expose step inputs _regardless_ of the stack they are running on (so all inputs are visible/usable on all stacks), having this overview of availability (as I mentioned above) can be very helpful. You don't have to peak into every input's description just to get the information of where it is available.

What do you think?

File: step.yml (hunk 3 above)
Comment (Bence1001) [reply step.yml:184]: Let's continue the discussion in that thread about this topic!

File: step.yml (hunk 4 above)
Comment (Bence1001) [reply step.yml:194]: Let's continue the discussion in that thread about this topic!

File: step.yml (hunk 8 above)
Comment (Bence1001) [reply step.yml:304]: Exactly, I just noticed this as well! Let's fix this in the next PR, as this step is going to change anyway.

File: step.yml (hunk 2 above)
Comment (BanyikAnna) [reply step.yml:166]: I like your reasoning, we can follow what works best for our users! ;)

File: step.yml (hunk 8 above)
Comment (BanyikAnna) [reply step.yml:304]: Thank you :)

File: step.yml (hunk 2 above)
Comment (Bence1001) [reply step.yml:166]: Cool!

Comment (ofalvai) [reply step.yml:166]: I would vote for the current highlight (in the title) because the description is only visible in the workflow editor if the user expands the summary by clicking on it. It's not the best UX and I think a lot of people would miss this important detail if it was visible only in the input description.

File: step.yml [hunk 10]
```diff
@@ -181,14 +181,25 @@ inputs:
...
+        Should be more than 1 if the Test Repetition Mode (`test_repetition_mode`) is other than `none`.
       is_required: true
+  - relaunch_tests_for_each_repetition: "no"
+    opts:
+      title: Relaunch Tests for Each Repetition (Available in Xcode 13+)
+      summary: If enabled, tests will launch in a new process for each repetition.
+      description: |-
+        If enabled, tests will launch in a new process for each repetition.
+
+        By default, tests launch in the same process for each repetition.
```
Comment (ofalvai) [inline step.yml:199]: Can we elaborate what it means to launch in the same process? Why it's useful in practice, what problem it solves, etc. I would also link the relevant documentation from Apple on the `xcodebuild` flag (if there's any)

**Review by ofalvai — APPROVED** [review]
LGTM, just a few minor documentation comments.

File: step.yml (hunk 10 above)
Comment (Bence1001) [reply step.yml:199]: Thanks for raising this!

`xcodebuild`'s command line documentation is not more verbose than our description; in fact, I almost copied it word-to-word:
```
$ xcodebuild -h
...
-test-repetition-relaunch-enabled YES|NO                 Enable or disable, tests repeating in a new process for each repetition.  Must be used in conjunction with -test-iterations, -retry-tests-on-failure, or -run-tests-until-failure. If not specified, tests will repeat in the same process.
```
`man xcodebuild` mentions [Xcode Builds Settings Reference](https://help.apple.com/xcode/mac/current/#/itcaec37c2a6), but it doesn't say anything about Test Repetition.

Would it make sense to talk more verbosely about this and its use case in DevCenter only?
